OP: preserve spatial semantic for Reshape

erizmr opened the pull request:
Hi @jackwish 
Here is the PR with the naive `fakeTranspose`. This naive version will add a `transpose` before every `Reshape`, which is definitely not our final target, a conditional statement is indeed needed. 

The condition that requires the additional `Transpose` layer is defined below as we discussed in #28 :
> if the input (or output) tensor of Reshape holds a special data layout semantic, we need to insert Transpose before (or after) the Reshape operator

I am wondering how to figure the condition out inside the `Reshape`. How do you think about it?

jackwish commented:
Maybe we can defer _inserting Transpose_ to `Reshape.transform()` where the layout propagation has been done - we know the input and output layout requirement if there were any. This looks better as the spatial semantic may also pass through _Transparent_ operators to `Reshape` and we don't need to worry about how the newly inserted `Transpose` should work together with layout propagation.

Btw, we adopt a _squash_ approach when merging PRs since we prefer to have a relatively clean commit history. In which way we rebase a _feature branch_ against _main branch_ sometimes.

jackwish commented on tests/test_ops.py:
```diff
@@ -40,6 +40,7 @@ def test_ops_implicit_layout():
         'maxpooling.float32',
         'resize-bilinear.float32',
         'resize-nearest-neighbor.float32',
+        'reshape-transpose.float32',
```
What about naming the file as something like `conv-reshape`? We are inserting `Transpose` to fix a `Conv` to `Reshape` issue here.

jackwish commented on tflite2onnx/op/reshape.py:
```diff
@@ -72,13 +73,45 @@ def parse(self):
 
         self.setParsed()
 
+    def fakeTranspose(self):
```
What about naming the function as something like `preserveInputSpatialSemantic`? We named `fakeBroadcast` in `Binary` because we were doing something like a _fake broadcasting_.

jackwish commented on tflite2onnx/op/reshape.py:
```diff
@@ -72,13 +73,45 @@ def parse(self):
+        # we need to insert Transpose before (or after) the Reshape operator
+        # https://github.com/jackwish/tflite2onnx/issues/28
+
+        # assert(self.status.initialized)
```
We can safely remove this or assert it's parsed.

erizmr commented:
Hi @jackwish , I have updated the `fakeTranspose`. As you suggested, I deferred the _inserting Transpose_ into `Reshape. transform()` and it will only be executed when the input layout is not None. I also added one more test model `reshape-transpose.float32.tflite` for the `fakeTranspose` case, though the Codecov is not fully satisfied with the new commits..

jackwish commented on tflite2onnx/op/reshape.py:
```diff
@@ -72,13 +73,45 @@ def parse(self):
+
+        transposed_name = 'TFLITE2ONNX_Transposed_%s' % to_transpose.name
+        transposed = self.TFactory.getWithRef(to_transpose, transposed_name, True)
+        transposed.shape = [to_transpose.shape[p] for p in [0, 2, 3, 1]]
```
The shape should be parsed from the `to_transpose.layout` and `to_transpose.shape` rather than assume it's an NHWC. You can get a hint from the [_layout transform_ utils](https://github.com/jackwish/tflite2onnx/blob/master/tflite2onnx/layout.py#L11). I think it should be ready to be called.

jackwish commented on tflite2onnx/op/reshape.py:
```diff
@@ -72,13 +73,45 @@ def parse(self):
+
+        # Construct the additional transpose
+        trans = Transpose(self.TFactory, -1)
+        trans.attrs['perm'] = np.array([0, 2, 3, 1])
```
Similar to shape handling above, we need to generalize this too.

jackwish commented on tflite2onnx/op/reshape.py:
```diff
@@ -72,13 +73,45 @@ def parse(self):
+        if i.layout is not None:
+            # Insert a `Transpose` if a required layout has propagated here
+            self.fakeTranspose()
         if not self.forFakeBroadcasting:
```
Maybe we need to reorder the logic here to something like
```py
# get comment var
if self.forFakeBroadcasting:
    ...
else
   ...

jackwish commented on tflite2onnx/op/reshape.py:
```diff
@@ -72,13 +73,45 @@ def parse(self):
 
     def transform(self):
+        i = self.inputs[0]
+        if i.layout is not None:
```
Could be `if i.layout`. I know I have some place that written like `if i.layout is not None`...

jackwish reviewed (changes requested):
@erizmr Thank you for your contribution!

I have several comments for the PR. In general, we can try to make our code generic so we don't need to rework for cases we have previously investigated but shifted a bit. And, could we have a similar test and code path for cases like `Reshape-Conv` for which case we may share some utils with?

jackwish commented:
Btw, I suggest resolving the heavy history in this PR on a side branch. You can always do `git rebase origin/master` (depending on how an upstream remote repo named for you) even if you are on the `master` branch. But ignore this if you are not sure how to do that.

And, don't worry about the CodeCov report, our target is to make code tested but not make CodeCov happy.

erizmr commented:
Hi @jackwish , thanks for your comments!
I have made the preserve spatial semantic part more generic as you suggested. 
Also, a `preserveOutputSpatialSemantic` function is added dealing with the `reshape-conv` cases. Though its structure is similar to the input one, I keep them separately temporarily for readability.  

Notably, one difference between the output and the input case is that sometimes there is more than one consumer for the output. There might be a chance for these consumers requiring different layouts, but I think it is a quite small chance for people to construct a network like that. So I assume the multiple consumers require the same spatial semantic currently. 

More test models are added, including `conv-reshape.float32`, 'conv-reshape-conv.float32', 'conv-reshape-multiple-conv.float32'.
One thing to mention: there is a `reshape-conv.float32` initially, but it has a NHWC input while a NCHW output layout, which seems not so convenient to be tested in the current framework. That's the reason for adding a `conv-reshape-conv` as a substitute.

Finally, sorry for the noisy heavy history.. I can give it a try to resolve.

erizmr commented:
BTW the tests `tests/test_quantize.py` in the build and test failed occasionally though they got passed well on my local machine. And it seems there is no modification for codes of quantize part, which is quite interesting.

jackwish commented on tests/test_ops.py:
```diff
@@ -40,6 +40,13 @@ def test_ops_implicit_layout():
+        'conv-reshape.float32',
+        'conv-reshape-conv.float32',
+        'conv-reshape-multiple-conv.float32',
+        # 'reshape-conv.float32',
```
So it's blocked by our test infra which tries to convert any 4D IO from NHWC to NCHW right? Maybe we can make `Reshape` have a 2D or 3D input.

erizmr replied:
Yeah exactly the dimension restrictions. Sorted by changing to 3D input. Thanks!

jackwish commented on tflite2onnx/op/reshape.py:
```diff
@@ -72,17 +72,143 @@ def parse(self):
+            shape_t.data = layout.transform(shape_t.data)
+        else:
+            if i.layout:
+                # Insert a `Transpose` before `Reshape`
```
We can merge the comment into one and put it right after `else` here.

jackwish commented on tflite2onnx/op/reshape.py:
```diff
@@ -72,17 +72,143 @@ def parse(self):
+        # Copy the layout of original input of `Reshape`
+        # e.g. current: NCHW, (source: NHWC, target: NCHW)
+        layout = copy.deepcopy(to_transpose.layout)
+        # Exchange the source and target layout for layout preservation
```
Hmmm, maybe we can construct a `Layout` object directly with the parameters from `to_transpose.layout`? And, `Layout.current` doesn't take significant currently if I remember correctly. So mostly we can ignore it.

jackwish commented on tflite2onnx/op/reshape.py:
```diff
@@ -72,17 +72,143 @@ def parse(self):
+        # Copy the layout of original output of `Reshape`
+        # e.g. current: NHWC, (source: NCHW, target: NHWC )
+        layout = copy.deepcopy(transposed.layout)
+        # Exchange the source and target layout for layout preservation
```
Similar to input here.

jackwish commented on tflite2onnx/op/reshape.py:
```diff
@@ -72,17 +72,143 @@ def parse(self):
+        assert(self.status.parsed)
+        transposed = self.outputs[0]
+
+        transposed_new_name = 'TFLITE2ONNX_Transposed_%s' % transposed.name
```
This seems to be a bit hacky, maybe we can just name the new tensor as something like `'TFLITE2ONNX_ToTransposed_%s' % transposed.name`?

jackwish reviewed (changes requested):
@erizmr Thank you for your productive work! We are very close to merging with several minor comments.

And, don't worry about the quantization test, we can rerun it to make it pass. Maybe I can relax the tolerance to low the failure rate.

erizmr commented:
Hi @jackwish 
The dimension restriction problem of `reshape-conv.float32` was sorted by turning to 3D input. Thanks for your suggestions!
Please also check the other changes.

As for the heavy history, I tried to squash them but git complains conflicts for the current changes in this PR, which I think it's a little bit dangerous to sort that from my remote branch. It will be great if you can help it when merging. Thanks a lot!

jackwish commented on tflite2onnx/op/reshape.py:
```diff
@@ -72,17 +73,135 @@ def parse(self):
+        to_transpose_name = transposed.name
+        # Rename the transposed tensor
+        # because it will be the output of the new `transpose`
+        self.TFactory.registery[transposed_new_name] = \
```
I mean for `[Reshape] -- <T_nhwc> -- [Conv]` in TFLite, after _layout propagation_ we got `[Reshape] -- <T_nchw> -- [Conv]` where we need to change it into `[Reshape] -- <T_nhwc> -- [Transpose] -- <T_nchw> -- [Conv]`. So we are building `<T_nhwc> -- [Transpose]` and insert it into the graph. We can simply name the `<T_nhwc>` as `ToTranspose`. The _to_ means this tensor is going to transpose.

Basically, the tensors and operators that start with a `TFLITE2ONNX` prefix are ones that we added to the graph as the helper, while others are converted from the TFLite model.

So maybe we can switch the name here in which way we don't need to modify the name of the existed tensors.

erizmr replied:
Ah, I gotchu. The motivation to use the hacky way here is to avoid the issue that two nodes have the same name. 
The operator is named after its first output tensor in the `setParsed` ( before the  _layout propagation_ I think). So the `Reshape` has its name after `<T_nchw>`. Then the newly inserted 'Transpose' will also be named after `<T_nchw>` when `setParsed`, which makes the `Reshape` and `Transpose` the same name. So either the output tensor of `Transpose` or the operator `Reshape` should be renamed.

According to the convention you have mentioned:

> Basically, the tensors and operators that start with a TFLITE2ONNX prefix are ones that we added to the graph as the helper, while others are converted from the TFLite model.

I think it might be better to rename `Transpose` directly (although still a bit hacky lol ) because it is also a helper operator. How do you think about it?

jackwish replied:
That's a good point, I don't even remember that we have used the name of the first tensor as node name. We didn't name the node at the beginning because TFLite operators are not named. But named operators do improve debug experience so we just named it with that policy which may require double consideration.

Renaming `Transpose` sounds good for me :)

jackwish commented on tflite2onnx/op/reshape.py:
```diff
@@ -72,17 +73,135 @@ def parse(self):
+        assert(self.status.parsed)
+        transposed = self.outputs[0]
+
+        transposed_new_name = 'TFLITE2ONNX_ToTransposed_%s' % transposed.name
```
Maybe the variable has a similar name with its content.

jackwish reviewed (changes requested):
@erizmr thanks! I think we can merge after a last minor change.

Don't worry about the commit history, they should be merged as one single commit eventually.

jackwish commented:
@erizmr Thank you for your contribution, this is now merged.

Btw it's somehow dangerous to open PR from master branch, it may be deleted unexpectedly after merging.

erizmr commented:
@jackwish 
Thanks for confirming this PR.
Yeah, indeed, will checkout to a new branch for the next PRs. 👍
